lib/storage/data/external/GCP/GcpClient.js

```diff
@@ -0,0 +1,522 @@
+     * @param {object} params - Contains Bucket name and VersioningConfiguration
+     * @param {function} callback
+     */
+    putBucketVersioning(params, callback) {
+        // GCP doesn't support S3-style bucket versioning
+        // This is a no-op to maintain API compatibility
+        callback(null, {});
+    }
+
+    /**
+     * List object versions in bucket.
+     * @param {object} params - Contains Bucket name and optional filters
+     * @param {function} callback
+     */
+    listVersions(params, callback) {
+        // GCP doesn't have S3-style versioning, but we can simulate this
+        // by returning current objects as "versions" for API compatibility
+        return this.listObjects(params , (err, data) => {
+            if (err) {
+                callback(err);
+                return;
+            }
+            const result = {
+                Name: data.Name,
+                Prefix: data.Prefix,
+                Delimiter: data.Delimiter,
+                    MaxKeys: data.MaxKeys,
+                    IsTruncated: data.IsTruncated,
+                    Versions: data.Contents ? data.Contents.map(obj => ({
+                        ...obj,
+                        VersionId: obj.ETag, // Use ETag as version ID
+                        IsLatest: true,
+                    })) : [],
+                    CommonPrefixes: data.CommonPrefixes,
+                };
+                callback(null, result);
+            });
+    }
+
+    /**
+     * Head an object.
+     */
+    headObject(params, callback) {
+        this.send(new HeadObjectCommand(params))
+            .then(data => callback(null, data))
+            .catch(err => callback(err));
+    }
+}
+
+function gcpSigningMiddleware(config) {
+    return (next, context) => async (args) => {
+        const { request } = args;
+        if (!request.headers) {
+            request.headers = {};
+        }
+        // Set x-goog-date if not present
+        if (!request.headers['x-goog-date']) {
+            request.headers['x-goog-date'] = new Date().toUTCString();
+        }
+        // Build string to sign
+        const fakeRequest = {
+            method: request.method,
+            headers: request.headers,
+            url: request.path,
+            path: request.path,
+            endpoint: { host: request.hostname },
+            virtualHostedBucket: undefined, // Set if needed
+            bucketName: config.bucketName,
+            query: request.query || {},
+        };
+        // If using virtual hosted style, set virtualHostedBucket
+        if (request.hostname && config.bucketName && request.hostname.startsWith(config.bucketName + '.')) {
+            fakeRequest.virtualHostedBucket = config.bucketName;
+            fakeRequest.gotBucketNameFromHost = true;
+        }
+        const data = {};
+        const logger = { trace: () => {} };
+        const stringToSign = constructStringToSignV2(fakeRequest, data, logger, 'GCP');
```

Comment:
I think we should probably use a real class logger here, and properly pass the request IDs, as here it would remove any logs in case of real issues?
(can be in a followup ticket?)

Reply:
up on this, just so we don't forget about the ticket